src/data_structures.rs

```diff
@@ -1,33 +1,58 @@
-use crate::{Rc, String, Vec};
+//!
+//! Contains traits for key material and evaluation proofs:
+//!  * for parameters and key material of the scheme: [`PCParameters`], [`PCVerifierKey`], [`PCCommitterKey`]
+//!  * for commitments and their hiding randomness: [`PCCommitment`], [`PCRandomness`]
+//!  * for a simple (single-polynomial single-point) opening proof: [`PCProof`]  
+//!  * for the [HaloInfinite] "batch evaluation" proof: [`PCMultiPointProof`]
+//!  * [`PCVerifierState`] as needed due to the verifier splitting.
+//!
+//! Additionally the auxiliary structures [`LabeledCommitment`], [`LabeledPolynomial`],
+//! [`LabeledRandomness`] are provided for internal usage.
+//!
+//! [HaloInfinite]: https://eprint.iacr.org/2020/1536
+use crate::{error::*, Rc, String};
 pub use algebra::DensePolynomial as Polynomial;
 use algebra::{
     serialize::{CanonicalDeserialize, CanonicalSerialize, SerializationError},
-    Field, SemanticallyValid, ToBytes,
+    Curve, Field, Group, SemanticallyValid, ToBytes, UniformRand,
 };
-use rand_core::RngCore;
 use std::{
-    borrow::Borrow,
     fmt::Debug,
     io::{Error as IoError, ErrorKind, Read, Result as IoResult, Write},
-    ops::{AddAssign, MulAssign, SubAssign},
 };
 
 /// Labels a `LabeledPolynomial` or a `LabeledCommitment`.
 pub type PolynomialLabel = String;
 
 /// Defines the minimal interface for public params for any polynomial
 /// commitment scheme.
-pub trait PCUniversalParams:
-    Clone + Debug + CanonicalSerialize + CanonicalDeserialize + Eq + PartialEq
+pub trait PCParameters<G: Curve>:
+    Clone + Debug + Eq + PartialEq + CanonicalSerialize + CanonicalDeserialize
 {
+    /// The committer key for the scheme; used to commit to a polynomial and then
+    /// open the commitment to produce an evaluation proof.
+    type CommitterKey: PCCommitterKey;
+    /// The verifier key for the scheme; used to check an evaluation proof.
+    type VerifierKey: PCVerifierKey;
+    /// The error type for the scheme.
+    type Error: std::error::Error + From<Error>;
+
     /// Outputs the maximum degree supported by the committer key.
     fn max_degree(&self) -> usize;
 
     /// Returns the hash of `self` instance.
     fn get_hash(&self) -> &[u8];
 
+    /// Specializes the public parameters for polynomials up to the given `supported_degree`
+    /// and for enforcing degree bounds in the range `1..=supported_degree`.
+    fn trim(
+        &self,
+        segment_size: usize,
+    ) -> Result<(Self::CommitterKey, Self::VerifierKey), Self::Error>;
```

Comment:
I would change the trim function to return a new instance of `PCParameters`, and rename the parameter `segment_size: usize` to `max_degree` or similar.

Reply:
Done